Based on [olivier's script](https://www.kaggle.com/ogrellier/xgb-classifier-upsampling-lb-0-283)

In [1]:
MAX_ROUNDS = 400
OPTIMIZE_ROUNDS = False
LEARNING_RATE = 0.07
EARLY_STOPPING_ROUNDS = 50  
# Note: I set EARLY_STOPPING_ROUNDS high so that (when OPTIMIZE_ROUNDS is set)
#       I will get lots of information to make my own judgment.  You should probably
#       reduce EARLY_STOPPING_ROUNDS if you want to do actual early stopping.

I recommend initially setting <code>MAX_ROUNDS</code> fairly high and using <code>OPTIMIZE_ROUNDS</code> to get an idea of the appropriate number of rounds (which, in my judgment, should be close to the maximum value of  <code>best_ntree_limit</code> among all folds, maybe even a bit higher if your model is adequately regularized...or alternatively, you could set <code>verbose=True</code> and look at the details to try to find a number of rounds that works well for all folds).  Then I would turn off <code>OPTIMIZE_ROUNDS</code> and set <code>MAX_ROUNDS</code> to the appropraite number of total rounds.  

The problem with "early stopping" by choosing the best round for each fold is that it overfits to the validation data.    It's therefore liable not to produce the optimal model for predicting test data, and if it's used to produce validation data for stacking/ensembling with other models, it would cause this one to have too much weight in the ensemble.  Another possibility (and the default for XGBoost, it seems) is to use the round where the early stop actually happens (with the lag that verifies lack of improvement) rather than the best round.  That solves the overfitting problem (provided the lag is long enough), but so far it doesn't seem to have helped.  (I got a worse validation score with 20-round early stopping per fold than with a constant number of rounds for all folds, so the early stopping actually seemed to underfit.)


In [2]:
import numpy as np
import pandas as pd
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from numba import jit
import time
import gc
import warnings
warnings.filterwarnings('ignore')

In [3]:
# Compute gini

# from CPMP's kernel https://www.kaggle.com/cpmpml/extremely-fast-gini-computation
@jit
def eval_gini(y_true, y_prob):
    y_true = np.asarray(y_true)
    y_true = y_true[np.argsort(y_prob)]
    ntrue = 0
    gini = 0
    delta = 0
    n = len(y_true)
    for i in range(n-1, -1, -1):
        y_i = y_true[i]
        ntrue += y_i
        gini += y_i * delta
        delta += 1 - y_i
    gini = 1 - 2 * gini / (ntrue * (n - ntrue))
    return gini

In [4]:
# Funcitons from olivier's kernel
# https://www.kaggle.com/ogrellier/xgb-classifier-upsampling-lb-0-283

# def gini_xgb(preds, dtrain):
#     labels = dtrain.get_label()
#     gini_score = -eval_gini(labels, preds)
#     return [('gini', gini_score)]

def target_encode(trn_series=None,    # Revised to encode validation series
                  val_series=None,
                  tst_series=None,
                  target=None,
                  min_samples_leaf=1,
                  smoothing=1,
                  noise_level=0):
    """
    Smoothing is computed like in the following paper by Daniele Micci-Barreca
    https://kaggle2.blob.core.windows.net/forum-message-attachments/225952/7441/high%20cardinality%20categoricals.pdf
    trn_series : training categorical feature as a pd.Series
    tst_series : test categorical feature as a pd.Series
    target : target data as a pd.Series
    min_samples_leaf (int) : minimum samples to take category average into account
    smoothing (int) : smoothing effect to balance categorical average vs prior
    """
    assert len(trn_series) == len(target)
    assert trn_series.name == tst_series.name
    temp = pd.concat([trn_series, target], axis=1)
    
    # Compute target mean
    averages = temp.groupby(by=trn_series.name)[target.name].agg(["mean", "count"])
    
    # Compute smoothing
    smoothing = 1 / (1 + np.exp(-(averages["count"] - min_samples_leaf) / smoothing))
    
    # Apply average function to all target data
    prior = target.mean()
    
    # The bigger the count the less full_avg is taken into account
    averages[target.name] = prior * (1 - smoothing) + averages["mean"] * smoothing
    averages.drop(["mean", "count"], axis=1, inplace=True)
    
    # Apply averages to trn and tst series
    ft_trn_series = pd.merge(
        left = trn_series.to_frame(trn_series.name),
        right = averages.reset_index().rename(columns={'index': target.name, target.name: 'average'}),
        on = trn_series.name, # index name
        how = 'left')['average'].rename(trn_series.name + '_mean').fillna(prior)
    
    # pd.merge does not keep the index so restore it
    ft_trn_series.index = trn_series.index # index copy
        # trn_series processing end
    
    ft_val_series = pd.merge(
        left = val_series.to_frame(val_series.name),
        right = averages.reset_index().rename(columns={'index': target.name, target.name: 'average'}),
        on = val_series.name,
        how = 'left')['average'].rename(trn_series.name + '_mean').fillna(prior)
    # pd.merge does not keep the index so restore it
    ft_val_series.index = val_series.index
        # val_series processing end
        
    ft_tst_series = pd.merge(
        left = tst_series.to_frame(tst_series.name),
        right = averages.reset_index().rename(columns={'index': target.name, target.name: 'average'}),
        on = tst_series.name,
        how = 'left')['average'].rename(trn_series.name + '_mean').fillna(prior)
    # pd.merge does not keep the index so restore it
    ft_tst_series.index = tst_series.index
        # tst_series processing end
        
    return add_noise(ft_trn_series, noise_level), add_noise(ft_val_series, noise_level), add_noise(ft_tst_series, noise_level)

def add_noise(series, noise_level):
    return series * (1 + noise_level * np.random.randn(len(series)))

In [5]:
# Read data
train_df = pd.read_csv('./input/train.csv', na_values="-1") # .iloc[0:200,:]
test_df = pd.read_csv('./input/test.csv', na_values="-1")

In [6]:
# from olivier
train_features = [
    "ps_car_13",  #            : 1571.65 / shadow  609.23
	"ps_reg_03",  #            : 1408.42 / shadow  511.15
	"ps_ind_05_cat",  #        : 1387.87 / shadow   84.72
	"ps_ind_03",  #            : 1219.47 / shadow  230.55
	"ps_ind_15",  #            :  922.18 / shadow  242.00
	"ps_reg_02",  #            :  920.65 / shadow  267.50
	"ps_car_14",  #            :  798.48 / shadow  549.58
	"ps_car_12",  #            :  731.93 / shadow  293.62
	"ps_car_01_cat",  #        :  698.07 / shadow  178.72
	"ps_car_07_cat",  #        :  694.53 / shadow   36.35
	"ps_ind_17_bin",  #        :  620.77 / shadow   23.15
	"ps_car_03_cat",  #        :  611.73 / shadow   50.67
	"ps_reg_01",  #            :  598.60 / shadow  178.57
	"ps_car_15",  #            :  593.35 / shadow  226.43
	"ps_ind_01",  #            :  547.32 / shadow  154.58
	"ps_ind_16_bin",  #        :  475.37 / shadow   34.17
	"ps_ind_07_bin",  #        :  435.28 / shadow   28.92
	"ps_car_06_cat",  #        :  398.02 / shadow  212.43
	"ps_car_04_cat",  #        :  376.87 / shadow   76.98
	"ps_ind_06_bin",  #        :  370.97 / shadow   36.13
	"ps_car_09_cat",  #        :  214.12 / shadow   81.38
	"ps_car_02_cat",  #        :  203.03 / shadow   26.67
	"ps_ind_02_cat",  #        :  189.47 / shadow   65.68
	"ps_car_11",  #            :  173.28 / shadow   76.45
	"ps_car_05_cat",  #        :  172.75 / shadow   62.92
	"ps_calc_09",  #           :  169.13 / shadow  129.72
	"ps_calc_05",  #           :  148.83 / shadow  120.68
	"ps_ind_08_bin",  #        :  140.73 / shadow   27.63
	"ps_car_08_cat",  #        :  120.87 / shadow   28.82
	"ps_ind_09_bin",  #        :  113.92 / shadow   27.05
	"ps_ind_04_cat",  #        :  107.27 / shadow   37.43
	"ps_ind_18_bin",  #        :   77.42 / shadow   25.97
	"ps_ind_12_bin",  #        :   39.67 / shadow   15.52
	"ps_ind_14",  #            :   37.37 / shadow   16.65
]
# add combinations
combs = [
    ('ps_reg_01', 'ps_car_02_cat'),  
    ('ps_reg_01', 'ps_car_04_cat'),
]

In [7]:
# Process data
id_test = test_df['id'].values
id_train = train_df['id'].values
y = train_df['target']

start = time.time()
for n_c, (f1, f2) in enumerate(combs):
    name1 = f1 + "_plus_" + f2
    print('current feature %60s %4d in %5.5f'
          % (name1, n_c + 1, (time.time() - start) / 60), end='')
    print('\r' * 75, end='')
    train_df[name1] = train_df[f1].apply(lambda x: str(x)) + "_" + train_df[f2].apply(lambda x: str(x))
    test_df[name1] = test_df[f1].apply(lambda x: str(x)) + "_" + test_df[f2].apply(lambda x: str(x))
    # Label Encode
    lbl = LabelEncoder()
    lbl.fit(list(train_df[name1].values) + list(test_df[name1].values))
    train_df[name1] = lbl.transform(list(train_df[name1].values))
    test_df[name1] = lbl.transform(list(test_df[name1].values))

    train_features.append(name1)
    
X = train_df[train_features]
test_df = test_df[train_features]

f_cats = [f for f in X.columns if "_cat" in f]
# catagory columns

current feature                                 ps_reg_01_plus_ps_car_04_cat    2 in 0.04339

In [8]:
y_valid_pred = 0*y

y_test_pred = 0
# initialization to predict

In [9]:
from collections import Counter
print('Counter : ',Counter(y_valid_pred))
print(y_valid_pred)

Counter :  Counter({0: 595212})
0         0
1         0
2         0
3         0
4         0
         ..
595207    0
595208    0
595209    0
595210    0
595211    0
Name: target, Length: 595212, dtype: int64


In [10]:
f_cats

['ps_ind_05_cat',
 'ps_car_01_cat',
 'ps_car_07_cat',
 'ps_car_03_cat',
 'ps_car_06_cat',
 'ps_car_04_cat',
 'ps_car_09_cat',
 'ps_car_02_cat',
 'ps_ind_02_cat',
 'ps_car_05_cat',
 'ps_car_08_cat',
 'ps_ind_04_cat',
 'ps_reg_01_plus_ps_car_02_cat',
 'ps_reg_01_plus_ps_car_04_cat']

In [11]:
# Set up folds
K = 5
kf = KFold(n_splits = K, random_state = 1, shuffle = True)
np.random.seed(0)

In [12]:
# Set up classifier
model = XGBClassifier(    
                        n_estimators=MAX_ROUNDS,
                        max_depth=4,
                        objective="binary:logistic",
                        learning_rate=LEARNING_RATE, 
                        subsample=.8,
                        min_child_weight=6,
                        colsample_bytree=.8,
                        scale_pos_weight=1.6,
                        gamma=10,
                        reg_alpha=8,
                        reg_lambda=1.3,
                     )

In [13]:
for i, (train_index, test_index) in enumerate(kf.split(train_df)):
    print(train_index)
    print(train_index.shape)
    print('\n')
    print(test_index)
    print(test_index.shape)
    print('-'*80)
    # kolded, shuffle = True

[     0      1      2 ... 595209 595210 595211]
(476169,)


[     3      8      9 ... 595201 595202 595204]
(119043,)
--------------------------------------------------------------------------------
[     0      1      3 ... 595208 595209 595210]
(476169,)


[     2      7     11 ... 595203 595205 595211]
(119043,)
--------------------------------------------------------------------------------
[     2      3      5 ... 595207 595209 595211]
(476170,)


[     0      1      4 ... 595200 595208 595210]
(119042,)
--------------------------------------------------------------------------------
[     0      1      2 ... 595209 595210 595211]
(476170,)


[    10     17     21 ... 595175 595192 595196]
(119042,)
--------------------------------------------------------------------------------
[     0      1      2 ... 595208 595210 595211]
(476170,)


[     5     12     16 ... 595206 595207 595209]
(119042,)
--------------------------------------------------------------------------------


In [14]:
# train_df['ps_reg_01'].mean()
# smoothing = 10
# smoothing = 1 / (1 + np.exp(-(train_df['ps_calc_01'].count() - 200) / smoothing))
# print(np.exp(-(train_df['ps_calc_01'].count() - 200)))


In [17]:
# Run CV

for i, (train_index, test_index) in enumerate(kf.split(train_df)):
    
    # Create data for this fold
    y_train, y_valid = y.iloc[train_index].copy(), y.iloc[test_index] # Tuple
    X_train, X_valid = X.iloc[train_index,:].copy(), X.iloc[test_index,:].copy() # pandas.Series
    X_test = test_df.copy()
    print( "\nFold ", i)
    
    # Encode data
    for f in f_cats:
        X_train[f + "_avg"], X_valid[f + "_avg"], X_test[f + "_avg"] = target_encode(
                                                        trn_series=X_train[f],
                                                        val_series=X_valid[f],
                                                        tst_series=X_test[f],
                                                        target=y_train,
                                                        min_samples_leaf=200,
                                                        smoothing=10,
                                                        noise_level=0
                                                        )
    print('Encode end')
    # Run model for this fold
        # OPTIMIZE_ROUNS is default == False
    if OPTIMIZE_ROUNDS:
        eval_set=[(X_valid,y_valid)]
        fit_model = model.fit( X_train, y_train, 
                               eval_set=eval_set,
                               eval_metric=gini_xgb,
                               early_stopping_rounds=EARLY_STOPPING_ROUNDS,
                               verbose=False
                             )
        print( "  Best N trees = ", model.best_ntree_limit )
        print( "  Best gini = ", model.best_score )
    else:
        print('fit start')
        fit_model = model.fit( X_train, y_train , verbose=True)
        print('model fit.')
        
    # Generate validation predictions for this fold
    pred = fit_model.predict_proba(X_valid)[:,1]
    print(pred)
    print( "  Gini = ", eval_gini(y_valid, pred) )
    y_valid_pred.iloc[test_index] = pred
    # Accumulate test set predictions
    y_test_pred += fit_model.predict_proba(X_test)[:,1]
        # y_test_pred < for submit.csv file
    del X_test, X_train, X_valid, y_train
    


Fold  0
Encode end
fit start
model fit.
[0.02368145 0.04678843 0.06983136 ... 0.0368156  0.03348101 0.05773815]
  Gini =  0.2849993810505098

Fold  1
Encode end
fit start
model fit.
[0.03058705 0.03388928 0.04379487 ... 0.09302994 0.05414506 0.02611043]
  Gini =  0.2819388953345532

Fold  2
Encode end
fit start
model fit.
[0.06399968 0.04613879 0.06331564 ... 0.03274471 0.06336704 0.03973275]
  Gini =  0.2763158671178396

Fold  3
Encode end
fit start
model fit.
[0.05342885 0.0789933  0.0460133  ... 0.0321997  0.0434684  0.09406571]
  Gini =  0.3022602735495421

Fold  4
Encode end
fit start
model fit.
[0.05312172 0.0323148  0.04120189 ... 0.05744714 0.03252183 0.0249834 ]
  Gini =  0.2840475033555092


In [64]:

y_test_pred /= K  # Average test set predictions

print( "\nGini for full training set:" )
eval_gini(y, y_valid_pred)

# Run CV end


Gini for full training set:


0.28568575853594425

In [62]:
def test_eval_gini(y_true, y_prob):
    y_true = np.asarray(y_true)
    print(Counter(y_true))
    print('type is',type(y_true))
    y_true = y_true[np.argsort(y_prob)]
        # Based on y_prob, return indexes in small order
    ntrue = 0
    gini = 0
    delta = 0
    n = len(y_true)
    print('len is ', n)
    for i in range(n-1, -1, -1):
        y_i = y_true[i]
        ntrue += y_i
        gini += y_i * delta
        delta += 1 - y_i
        # if y value is contain '1',
            # gini value is improve ++delta values
            # and delta value is retained
        # if y value is contain '0',
            # delta values improve ++1,
    
    print('true count : ', ntrue)
    print('gini is : ', gini)
    # print('delta value is change : ', delta)
    print('ntrue * (n-nture) is : ',ntrue*(n-ntrue))
    gini = 1 - (2 * gini) / (ntrue * (n - ntrue))
    # gini = 1- ()
    return gini

In [63]:
test_eval_gini(y, y_valid_pred)

Counter({0: 573518, 1: 21694})
type is <class 'numpy.ndarray'>
len is  595212
true count :  21694
gini is :  4443712999
ntrue * (n-nture) is :  12441899492


0.28568575853594425

In [65]:
Counter(y)

Counter({0: 573518, 1: 21694})

In [68]:
# Save validation predictions for stacking/ensembling
val = pd.DataFrame()
val['id'] = id_train
val['target'] = y_valid_pred.values
val.to_csv('./result/xgb_valid.csv', float_format='%.6f', index=False)

In [69]:
# Create submission file
sub = pd.DataFrame()
sub['id'] = id_test
sub['target'] = y_test_pred
sub.to_csv('./result/xgb_submit.csv', float_format='%.6f', index=False)

Notes:<br>
version 16. Baseline best CV=.2832, LB=.282<br>
version 15. Ntree optimization for baseline<br>
version 21. Verbose version of baseline optimization<br>
version 22. Baseline + per-fold early stopping after 20 rounds<br>
version 23. Back to baseline.<br>
version 24. Some parameter tuning.<br>
version 25. Re-published to make it visible.<br>
version 26. A little more tuning.<br>
version 27: More tuning, get rid of upsampling (using  **<code>scale_pos_weight</code>** instead),<br>
                    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
                    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
                    Set <code>OPTIMIZE_ROUNDS</code> and <code>verbose</code> temporarily<br>
version 28: <code>MAX_ROUNDS=300</code> as a compromise<br>
version 29: Substantively identical. (Turn off now-irrelevant <code>verbose</code>.)<br>
version 30: Still substantively identical. Some visual cleanup.<br>
version 35. More tuning. CV went up but LB sorts lower (still .283)<br>
version 36. Identical (except turn off irrelevant <code>verbose</code>). Republished to make it visible.<br>
versions 37-42. More tuning (gamma=10, alpha=8). LB .284 (\*end zone dance\*).<br>
version 43. More tuning (min_child_weight=6).  LB score has considerably improved according to sort, but still .284